### PR TITLE
feat(documents): aggregate api

### PR DESCRIPTION
**Pull request by andersfylling**
This includes a change to BaseResourceAPI to enable the use of a operation-specific item type for aggregates, which is not the same as the REST resource. This follows our API design guide that states that aggregations can use pagination, which we use in the documents API. In the documents API only the `allUniqueValues` supports pagination.

### CODE REVIEW

File: packages/stable/src/__tests__/api/documentsApi.int.spec.ts [hunk 1]
```diff
@@ -215,4 +218,30 @@ describe('Documents integration test', () => {
...
+    const resp = await client.documents.aggregate({
+      filter: {
+        equals: {
+          property: ['type'],
+          value: 'PDF',
+        },
+      },
+      aggregate: 'count',
+    });
+    expect(resp.items).toBeGreaterThanOrEqual(1);
```
Comment (henrist) [inline packages/stable/src/__tests__/api/documentsApi.int.spec.ts]:
```suggestion
    expect(resp.items).toHaveLength(1);
```

File: packages/stable/src/__tests__/api/documentsApi.int.spec.ts [hunk 2]
```diff
@@ -215,4 +218,30 @@ describe('Documents integration test', () => {
...
+    expect(resp.items[0].count).toBeGreaterThanOrEqual(1);
+  });
+
+  test('document aggregate unique', async () => {
+    const resp =
+      await client.documents.aggregate<DocumentsAggregateUniqueValuesResponse>({
+        aggregate: 'uniqueValues',
+        properties: [{ property: ['mimeType'] }],
+      });
+    expect(resp.items).toBeGreaterThanOrEqual(0);
```
Comment (henrist) [inline packages/stable/src/__tests__/api/documentsApi.int.spec.ts]:
```suggestion
    expect(resp.items.length).toBeGreaterThanOrEqual(1);
```

File: packages/stable/src/__tests__/api/documentsApi.int.spec.ts [hunk 3]
```diff
@@ -215,4 +218,30 @@ describe('Documents integration test', () => {
...
+        },
+      },
+      aggregate: 'count',
+    });
+    expect(resp.items).toBeGreaterThanOrEqual(1);
+    expect(resp.items[0].count).toBeDefined();
+    expect(resp.items[0].count).toBeGreaterThanOrEqual(1);
+  });
+
+  test('document aggregate unique', async () => {
```
Comment (henrist) [inline packages/stable/src/__tests__/api/documentsApi.int.spec.ts]:
```suggestion
  test('document aggregate uniqueValues', async () => {
```

File: packages/stable/src/__tests__/api/documentsApi.unit.spec.ts [hunk 4]
```diff
@@ -385,4 +385,52 @@ describe('Documents unit test', () => {
...
+          value: 'PDF',
+        },
+      },
+      aggregate: 'count',
+    });
+    expect(resp.items).toHaveLength(1);
+    expect(resp.items[0]).toEqual({ count: 3456 });
+  });
+
+  test('document aggregate unique', async () => {
```
Comment (henrist) [inline packages/stable/src/__tests__/api/documentsApi.unit.spec.ts]:
```suggestion
  test('document aggregate uniqueValues', async () => {
```

File: packages/stable/src/api/documents/documentsApi.ts [hunk 5]
```diff
@@ -65,6 +67,27 @@ export class DocumentsAPI extends BaseResourceAPI<Document> {
     return this.documentContent(id);
   };
 
+  /**
+   * []()
```
Comment (henrist) [inline packages/stable/src/api/documents/documentsApi.ts]: ?

File: packages/stable/src/api/documents/types.gen.ts [hunk 6]
```diff
@@ -1,9 +1,6 @@
 // Do not modify this file!
-// It has been automatically generated and should not be edited manually
-// If you need to make changes to this file, please reach out to tommy.thorsen@cognite.com
-// or anders.fylling@cognite.com
-
-import { CogniteInternalId, CogniteExternalId } from '@cognite/sdk-core';
+// It was generated by the command "yarn generate-types stable:documents".
+// Instead update the code generation logic or the service contracts.
```
Comment (henrist) [inline packages/stable/src/api/documents/types.gen.ts]: This should not be changed (yet)?

**Review by henrist — COMMENTED** [review]
Should there also be some tests for allUniqueValues? Do we want to actually fetch all pages for the user in this case, or otherwise see if we can use the generator pattern (?) used elsewise for this?

**Comment by andersfylling** [comment]
> Do we want to actually fetch all pages for the user in this case, or otherwise see if we can use the generator pattern (?) used elsewise for this?

@henrist 
Not sure if we can use the generator here. BaseResource#listEndpoint utilizes makeAutoPaginationMethods. However, the BaseResource is current configured to use a Document[] as it's generic type. This fits the need for /search and /list, not so much for aggregate.

It would work better if aggregate was moved to it's own resource base inherited class, perhaps we can try that. The SDK would instead look like:
```ts
documents.aggregate.count(..)
documents.aggregate.uniqueValues(..)
documents.aggregate.allUniqueValues(..)
```

I split these into separate methods as each case uses different request and response structures, so it seems natural. Perhaps this should have been 3 sub-endpoints (`/documents/aggregate/count`, etc.) instead of using a constant in the request payload?

**Comment by henrist** [comment]
@andersfylling 

I support the idea of trying out a subresources in the SDK, and like the idea of different methods.

Regarding the API for the service I think it's fine to stay as-is, and I'd probably do the same again. In my head we have a separation between low level and high level abstractions here. The API to the service is low level. It changes infrequent and is pretty resistant against changes. It's more about passing operations with data around, and in thise case REST-ish calls. It's difficult to be (and stay) precise in this context. This SDK is more like a mix between both, and in this case it would be a higher level abstraction since we don't map 1-1 on the API and have additional behaviour such as the auto-pagination.

File: packages/core/src/baseResourceApi.ts [hunk 7]
```diff
@@ -312,11 +322,11 @@ export abstract class BaseResourceAPI<ResponseType> {
       .reduce((a, b) => [...a, ...b], []);
   }
 
-  protected addNextPageFunction<QueryType extends FilterQuery>(
-    endpoint: ListEndpoint<QueryType, ResponseType[]>,
-    cursorResponse: CursorResponse<ResponseType[]>,
+  protected addNextPageFunction<QueryType extends FilterQuery, Item>(
```
Comment (andersfylling) [inline packages/core/src/baseResourceApi.ts:325]: Item should perhaps default to ResponseType, thought it doesn't seem to have mattered?

File: packages/stable/src/api/documents/aggregateApi.ts [hunk 8]
```diff
@@ -1,53 +1,78 @@
...
+      aggregate: 'allUniqueValues',
+    });
   };
 
-  private async aggregate<
-    Request extends DocumentsAggregateRequest,
-    Response extends DocumentsAggregateResponse
-  >(request: Request): Promise<Response> {
-    const response = await this.post<Response>(this.url(), {
+  protected callAggregateCursorEndpointWithPost = async <
```
Comment (andersfylling) [inline packages/stable/src/api/documents/aggregateApi.ts]: private

**Review by henrist — APPROVED** [review]

**Comment by henrist** [comment]
@cognitedata/developer-tooling Can you please help review this?

**Comment by henrist** [comment]
Last failures far are due to flakes or issues during integration tests. Will trigger another rerun later if it persists.

**Review by cognite-ornellas — APPROVED** [review]

**Pull request merged**